cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java

```diff
@@ -289,34 +259,12 @@ protected void updateJobQueue(DatasetContext context, Notification notification)
       }
       DatasetId datasetId = DatasetId.fromString(datasetIdString);
       for (ProgramSchedule schedule : getSchedules(context, Schedulers.triggerKeyForPartition(datasetId))) {
-        JobQueueDataset jobQueue = getJobQueue(context);
-        boolean existingJobModified = false;
-        try (CloseableIterator<Job> jobs = jobQueue.getJobsForSchedule(schedule.getScheduleId())) {
-          while (jobs.hasNext()) {
-            Job job = jobs.next();
-            if (job.getState() != Job.State.PENDING_LAUNCH) {
-              // check if its PENDING_LAUNCH. If so, create a new Job to avoid the chance that the job-launching
-              // process has conflict
-              jobQueue.addNotification(job, notification);
-              existingJobModified = true;
-            }
-          }
-        }
-        // if no existing job was modified with the new notification, add a new job with the first notification
-        if (!existingJobModified) {
-          Job job = new SimpleJob(schedule, System.currentTimeMillis(),
-                                  Lists.newArrayList(notification), Job.State.PENDING_TRIGGER);
-          jobQueue.put(job);
-        }
+        JobQueueDataset jobQueue = Schedulers.getJobQueue(context, datasetFramework);
```

Comment:
you must use the same instance of JobQueueDataset as other methods in the same transaction. Otherwise you risk them overwriting each other's data without conflict. Please use a dataset cache